server/controllers/auth/login.js

```diff
@@ -1,3 +1,31 @@
-exports.login = (req, res) => {
-  res.send(' Hello from login file');
+const { compare } = require('bcrypt');
+const { jwtSign } = require('../../helpers');
+const { getUserByUsername } = require('../../database/queries/users');
+
+exports.login = (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) next({ code: 400, msg: 'bad request !!!!' });
+
+  const key = process.env.KEY;
+  let id;
+  getUserByUsername(username)
+    .then(({ rows }) => {
+      id = rows[0].id;
+      return compare(password, rows[0].password);
+    })
+    .then((isValid) => {
+      if (isValid) {
+        return jwtSign({ userInfo: { username, id } }, key);
+      } throw new Error(' username or password doesn\'t match our records ');
+    }).then((token) => {
+      res.cookie('token', token);
```

Comment:
how/where is the token going to be verified? You should set `maxAge` and `httpOnly` on the cookie.

I would however strongly suggest, instead of doing all of this yourselves, which is very easy to get wrong, use for example [this](https://expressjs.com/en/resources/middleware/cookie-session.html). You can get rid of the token signing and checking, instead you configure it as a middleware during server startup and then here you just do:
```
if (isValid) {
    var user = { 
        id, 
        username 
    };
    req.session.user = user;
}
```
And that's it! Then whenever you want to verify that a user is logged in, you do `if (req.session.user) {`
And lastly, to log out a user, you do `req.session = null`

Reply:
good point I wrote a middle-ware in ```PR #38```  to do this but her I would briefer to send a cookie if I need it in the front-end

Reply:
the cookie is handled automatically by the browser, you don't typically deal with it in the front end. Plus the solution that I proposed above does the same thing (it creates a cookie), it just hides the implementation so you don't have to deal with details like creating and verifying the token, etc. It's very easy to set up:
```
app.use(cookieSession({
  name: 'session',
  keys: ['key1']
}))
```
and then you can use it as I described above. Check out the link I posted in the first comment for examples.
I'm pushing you towards it cause it's gonna save time if you do it now rather then leaving it for later and then refactoring it.. :)

Reply:
ok nevermind, for some reason I thought the cookie-session middleware is using JWTs in the background, but apparently not, although it works similarly in principle. But let's go with the JWTs then.